### ISSUE TITLE
Replace `erase_memory()` with more performant alternative

### ISSUE THREAD

**Issue by dimakuv** [gramineproject/gramine#1257]
### Description of the feature

Gramine has `erase_memory()` function: https://github.com/gramineproject/gramine/blob/0119fd23ebbc892b33868e9ea9296ba6e8061685/common/include/api.h#L403-L410

This function was borrowed from Intel SGX SDK, but it is **very slow**.

Currently, it is mainly used in Protected Files implementation, in a performance-critical path of generating new per-block encryption keys (this happens on every 4KB write): https://github.com/gramineproject/gramine/search?q=erase_memory

I argue that we shouldn't use this function at all, and instead replace it with more performant alternatives. For example, mbedTLS has this: https://github.com/Mbed-TLS/mbedtls/blob/v3.3.0/library/platform_util.c#L39-L73

Since mbedTLS's implementation uses `memset()` (in a way that is confusing to compilers, so that they don't optimize it away), it is much more performant than our `erase_memory()` func.

### Why Gramine should implement it?

For performance reasons.

**Comment by kailun-qin** [comment]
Looks reasonable to me. I can work on this (if we make the decision to replace it eventually). Thanks!

**Comment by mkow** [comment]
Why not implement it like all other memset/memcpy/etc we implement, i.e. in assembly?

> in a way that is confusing to compilers, so that they don't optimize it away

I don't like it, it's one compiler update away from being silently removed when the optimizer gets better. And we will never notice that this happened.

Anyways, I don't care about this memory erasing, we have to have the live encryption keys in the memory anyways, so IMO this protection doesn't really change much. Just please don't accidentally make it misleading (i.e. present in the source but disappearing after compilation).

**Comment by dimakuv** [comment]
> Why not implement it like all other memset/memcpy/etc we implement, i.e. in assembly?

We could, but how can we prevent this assembly from being optimized away too?

**Comment by kailun-qin** [comment]
> Optimizer runs only on C code and intermediate representations, assembly is emitted as-is to the binary (or rather: copy pasted literally into the final assembly source which gets compiled into the binary).

Yep, but aren't we implementing `memset()` in (inline) assembly already for x86_64? I assume it won't be optimized away by the compiler.
https://github.com/gramineproject/gramine/blob/25ddc0effbf84c4a6b1a2d4bce93f3113f1ec04b/common/src/string/memcpy.c#L28

Well, if we consider cases other than x86_64 (written in C in our case), normally we have to do similar tricks like the volatile function pointer (e.g., used by mbedTLS or NetBSD) or `__asm__ __volatile__ ("" : : "r"(d) : "memory")` after `memset()` (e.g., used by musl).

**Comment by dimakuv** [comment]
> Optimizer runs only on C code and intermediate representations, assembly is emitted as-is to the binary (or rather: copy pasted literally into the final assembly source which gets compiled into the binary).

Ah, that's interesting, turns out I understood LTO (Link Time Optimization) completely wrong all this time.

I used to think of LTO as "binary-to-binary" optimization (that is, the LTO linker takes `.o` assembly files, lumps them together, runs optimizations and emits the final optimized assembly file). In other words, I thought that assembly *can be optimized*.

But turns out that LTO is a "intermediate-representation-to-binary" optimization (that is, the compiler produces intermediate representation, though confusingly put in `.o` files, and the LTO linker lumps the *intermediate representations* together, runs the optimizations on the intermediate representation, and then emits the final assembly).

See also https://johnnysswlab.com/link-time-optimizations-new-way-to-do-compiler-optimizations/, section "Under the hood".

---

TLDR: Yeah, we should just use inline asm in all such cases where we don't want any possiblity of optimizations.

**Comment by dimakuv** [comment]
I agree with @kailun-qin that we already have `memset()` implemented via inline assembly, so it won't be optimized away. And we currently don't support anything other than x86-64, so I wouldn't care about other archs.

**Comment by mkow** [comment]
> turns out I understood LTO (Link Time Optimization) completely wrong all this time

Heh, yes, seems so :)

> I agree with @kailun-qin that we already have `memset()` implemented via inline assembly, so it won't be optimized away.

That's not true, it _will_ be optimized away. Proof: https://godbolt.org/z/5vKKqKabf.
`memset` is a special name in C and C++ and the compiler is allowed to assume its semantics when optimizing code. It stops to be removed if you rename `memset` to e.g. `asdfmemset` in the linked example.

**Comment by dimakuv** [comment]
> That's not true, it will be optimized away. Proof: https://godbolt.org/z/5vKKqKabf.

Oh my. Compilers got way too aggressive. I didn't know that.

**Comment by kailun-qin** [comment]
> That's not true, it will be optimized away. Proof: https://godbolt.org/z/5vKKqKabf.
memset is a special name in C and C++ and the compiler is allowed to assume its semantics when optimizing code. It stops to be removed if you rename memset to e.g. asdfmemset in the linked example.

Interesting! Looks like the solution here is to implement `erase_memory()` in inline asm (i.e., to simply copy-paste the part of  x86-64 in our `memset()`) or asm (i.e., to adapt from another memset implementation).
